### PR TITLE
[FocusMode] Add title bar domain label browser tests

### PR DESCRIPTION
**Pull request by thypon**
These browser tests verify that the Focus Mode title bar's domain label keeps tracking the active page at the Brave presentation layer (`FocusModeTitleBarView::Update`), which previously had no coverage for these navigation paths:

- **ClientRedirectUpdatesDomain** — a client-side redirect must not leave the label on the original host
- **ServerRedirectUpdatesDomain** — a server-side redirect must update the label to the destination host
- **BackNavigationUpdatesDomain** — going back must restore the previous page's host in the label

A stale label would misrepresent the page being shown, so the tests drive the navigation types where that could regress. The title bar's own content formatting is covered by unit tests.

All 12 Focus Mode browsertests pass locally (9 existing + 3 new).

Requested in the security review of #38242 (brave/reviews#2215). Does not auto-close that PR.

### CODE REVIEW

File: browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+// prerender path additionally rejects the cross-site host transition
+// (a.test -> b.test) that would make this test meaningful. Upstream
+// tab_ui_helper_browsertest.cc already covers TabUIHelper's reaction
+// (ShouldNotAffectTabUIHelperOnPrerendering), so the Brave presentation layer
+// adds nothing testable here.
+
+// Restoring a page from the back/forward cache must update the domain label
+// back to the restored page's host. bfcache restore fires DidFinishNavigation;
+// TabUIHelper notifies, and the title bar must re-render with the prior host.
+IN_PROC_BROWSER_TEST_F(FocusModeUrlCoherenceBrowserTest,
```
Comment (netzenbot) [inline browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: This test never verifies the bfcache path it's named after: the fixture only enables `kBraveFocusMode`, and nothing asserts the restore came from the cache — the comment on the `GoBack()` even concedes it may be "a normal back navigation". Upstream bfcache tests enable it explicitly via `content::GetDefaultEnabledBackForwardCacheFeaturesForTesting()` and then assert `rfh->GetLifecycleState() == RenderFrameHost::LifecycleState::kInBackForwardCache` before going back (see `chrome/browser/ui/views/location_bar/location_bar_view_browsertest.cc`, `LocationBarViewGeolocationBackForwardCacheBrowserTest`). Either follow that pattern or rename the test to reflect that it covers back navigation. ([✅ Verify Tests Actually Test What They Claim](https://github.com/brave/brave-core/tree/master/docs/best-practices/testing-isolation.md#TI-012))

File: browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+#include "brave/browser/ui/focus_mode/focus_mode_features.h"
+#include "brave/browser/ui/views/frame/brave_browser_view.h"
+#include "brave/browser/ui/views/frame/focus_mode_title_bar_view.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_window/public/browser_window_features.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "chrome/browser/ui/views/frame/browser_view.h"
+#include "chrome/test/base/in_process_browser_test.h"
+#include "chrome/test/base/ui_test_utils.h"
+#include "content/public/browser/web_contents.h"
```
Comment (netzenbot) [inline browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: nit: `web_contents->GetController().GoBack()` needs the complete `content::NavigationController` type, but `content/public/browser/web_contents.h` only forward-declares it (see the `class NavigationController;` declaration there). Please add `#include "content/public/browser/navigation_controller.h"` rather than relying on a transitive include. ([✅ Always Include What You Use (IWYU)](https://github.com/brave/brave-core/tree/master/docs/best-practices/coding-standards.md#CS-001))

File: browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  // /server-redirect?<url> responds with HTTP MOVED PERMANENTLY to <url>.
+  const GURL redirect_url = https_server_.GetURL(
+      "a.test", "/server-redirect?" +
+                    https_server_.GetURL("b.test", "/empty.html").spec());
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), redirect_url));
+
+  EXPECT_TRUE(WaitForDomainToContain(u"b.test"));
+}
+
+// Activating a prerendered page would be the fourth edge case, but Prerender2
```
Comment (netzenbot) [inline browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: nit: this block documents a test that doesn't exist and reads as authoring notes ("would be the fourth edge case") rather than something a future reader of this file needs. Suggest dropping it, or trimming to a one-liner if the prerender caveat is genuinely worth preserving. ([✅ Comments Must Make Sense to Future Readers of the Codebase](https://github.com/brave/brave-core/tree/master/docs/best-practices/coding-standards.md#CS-019))

File: browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+  // /client-redirect?<url> returns an HTML page that meta-refreshes to <url>.
+  const GURL redirect_url = https_server_.GetURL(
+      "a.test", "/client-redirect?" +
+                    https_server_.GetURL("b.test", "/empty.html").spec());
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), redirect_url));
+
+  // The redirect is client-initiated, so wait for the destination host to
+  // appear in the label rather than asserting synchronously.
+  EXPECT_TRUE(WaitForDomainToContain(u"b.test"));
```
Comment (netzenbot) [inline browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: nit: on failure these only print `false`, which makes a CI timeout hard to diagnose. Consider `<< domain_text()` on the EXPECT_TRUE/ASSERT_TRUE calls (or a `SCOPED_TRACE` inside `WaitForDomainToContain`) so the label's actual contents show up in the output. ([✅ Include Diagnostic State in Test Assertions](https://github.com/brave/brave-core/tree/master/docs/best-practices/testing-isolation.md#TI-014))

File: browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+#include "ui/views/test/views_test_utils.h"
+#include "url/gurl.h"
+
+// Verifies that the Focus Mode title bar's domain label stays coherent with the
+// active page across navigation edge cases (client/server redirects and
+// back/forward cache restore). The title bar subscribes to TabUIHelper change
+// callbacks; these tests exercise the presentation layer
+// (FocusModeTitleBarView::Update) rather than the data source alone, which is
+// already covered by tab_ui_helper_browsertest.cc upstream.
+class FocusModeUrlCoherenceBrowserTest : public InProcessBrowserTest {
```
Comment (netzenbot) [inline browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: nit: this fixture is a verbatim copy of `FocusModeTitleBarViewBrowserTest` in `focus_mode_title_bar_view_browsertest.cc` (same helpers, feature list, server setup, animation scope). Consider adding these tests to that file or extracting the shared fixture into a small test header so the helpers only exist once. ([✅ Extract Repetitive Test Assertions into Helpers](https://github.com/brave/brave-core/tree/master/docs/best-practices/testing-isolation.md#TI-024))

**Review by zenparsing — COMMENTED** [review]
Let's add to the existing `focus_mode_title_bar_view_browsertest.cc` rather than creating a new browser test file.

File: browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc (hunk 5 above)
Comment (bridiver) [reply browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: We already have FocusModeTitleBarViewBrowserTest, but most of this looks like it could be unit tested?

**Comment by thypon** [comment]
Addressed review feedback in 020d3a5e9ea:

- @zenparsing / @bridiver: moved the three URL-coherence tests into the existing `focus_mode_title_bar_view_browsertest.cc` as a `FocusModeUrlCoherenceBrowserTest : public FocusModeTitleBarViewBrowserTest` subclass, reusing the existing `https_server_` / `browser_view()` / `title_bar()` / `domain_text()` / `WaitForDomainToContain` helpers. Deleted the standalone `focus_mode_url_coherence_browsertest.cc` and its BUILD.gn entry. (Re: unit-testing — these exercise the presentation layer via TabUIHelper change callbacks across navigations, which is already covered upstream at the data-source level by `tab_ui_helper_browsertest.cc`; happy to revisit if you'd prefer a unit-test approach instead.)
- @netzenbot bfcache: the subclass now enables `content::GetDefaultEnabledBackForwardCacheFeaturesForTesting()` and the restore test asserts `cached_rfh->GetLifecycleState() == content::RenderFrameHost::LifecycleState::kInBackForwardCache` before `GoBack`, so the cached-RFH path is verified rather than a possible normal back navigation.
- @netzenbot IWYU: added `#include "content/public/browser/navigation_controller.h"` (plus `render_frame_host.h`, `web_contents.h`, `back_forward_cache_util.h`, `browser_test_utils.h`, `default_handlers.h`, `utf_string_conversions.h` needed by the moved code).
- @netzenbot prerender comment: dropped the "would be the fourth edge case" block.
- @netzenbot diagnostics: appended `<< base::UTF16ToUTF8(domain_text())` to the `WaitForDomainToContain` assertions in the moved tests so timeouts print the current label.

Format and presubmit pass locally; `gn check` is blocked by a pre-existing missing `chrome/browser/new_tab_page/prefs/BUILD.gn` on this checkout (reproduces on a clean tree).

File: browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc [hunk 6]
```diff
@@ -119,3 +130,120 @@ IN_PROC_BROWSER_TEST_F(FocusModeTitleBarViewBrowserTest, TitleBarWiring) {
...
+      GetWebContents()->GetPrimaryMainFrame();
+
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(
+      browser(), https_server_.GetURL("b.test", "/empty.html")));
+  ASSERT_TRUE(WaitForDomainToContain(u"b.test"))
+      << base::UTF16ToUTF8(domain_text());
+
+  // Verify the previous page was actually cached, not evicted to a normal
+  // back navigation. Without this assertion GoBack could pass via either path.
+  ASSERT_EQ(cached_rfh->GetLifecycleState(),
```
Comment (netzenbot) [inline browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc]: `cached_rfh` is a bare `RenderFrameHost*`, so if the page is *not* bfcached it gets destroyed during the b.test navigation and this `ASSERT_EQ` dereferences a dangling pointer instead of failing cleanly. The upstream test this follows (`LocationBarViewGeolocationBackForwardCacheBrowserTest.CheckGeolocationIconVisibility` in `chrome/browser/ui/views/location_bar/location_bar_view_browsertest.cc`) holds a `content::RenderFrameDeletedObserver` and checks `deleted()` before reading `GetLifecycleState()`; `content::RenderFrameHostWrapper` (`content/public/test/test_utils.h`) wraps the same guard and is a near drop-in here. ([✅ Search for Existing Chromium Patterns](https://github.com/brave/brave-core/tree/master/docs/best-practices/testing-isolation.md#TI-006))

File: browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc [hunk 7]
```diff
@@ -119,3 +130,120 @@ IN_PROC_BROWSER_TEST_F(FocusModeTitleBarViewBrowserTest, TitleBarWiring) {
   browser()->GetFeatures().focus_mode_controller()->SetEnabled(false);
   EXPECT_FALSE(title_bar()->GetVisible());
 }
+
+// Verifies that the Focus Mode title bar's domain label stays coherent with the
+// active page across navigation edge cases (client/server redirects and
+// back/forward cache restore). The title bar subscribes to TabUIHelper change
+// callbacks; these tests exercise the presentation layer
+// (FocusModeTitleBarView::Update) rather than the data source alone, which is
+// already covered by tab_ui_helper_browsertest.cc upstream.
```
Comment (netzenbot) [inline browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc]: nit: the trailing clause ("rather than the data source alone, which is already covered by tab_ui_helper_browsertest.cc upstream") reads as a reply to the review discussion rather than something a fresh reader of this file needs. Suggest trimming to what the tests actually cover and dropping the justification against the alternative. ([✅ Comments Must Make Sense to Future Readers of the Codebase](https://github.com/brave/brave-core/tree/master/docs/best-practices/coding-standards.md#CS-019))

**Comment by thypon** [comment]
Addressed the latest netzenbot comments in 55ae3087eb4:

- L238 dangling `RenderFrameHost*`: wrapped `cached_rfh` in `content::RenderFrameHostWrapper` (`content/public/test/test_utils.h`) and added `ASSERT_FALSE(cached_rfh.IsRenderFrameDeleted())` before reading `GetLifecycleState()`, so an evicted page fails cleanly instead of dereferencing a dangling pointer. Matches the upstream `LocationBarViewGeolocationBackForwardCacheBrowserTest` pattern.
- L139 comment: trimmed the trailing "rather than the data source alone, which is already covered by tab_ui_helper_browsertest.cc upstream" clause.

Format + presubmit pass; all 3 tests green locally.

**Review by zenparsing — APPROVED** [review]
Looks good.

Although, it's not entirely clear to me that any paths are being covered here that aren't already covered in the title bar unit test. These tests seem to be testing the TabUIHelper more than the title bar. Did you take a look at that unit test?

**Comment by thypon** [comment]
> Looks good.
>
> Although, it's not entirely clear to me that any paths are being covered here that aren't already covered in the title bar unit test. These tests seem to be testing the TabUIHelper more than the title bar. Did you take a look at that unit test?

For anything that changes critical components such as urlbar visualization I feel we need to cover full flow/wiring, instead of unit testing the single components.

If anything changes in the wiring, and we don't test the full flow, we might introduce `brave-core` specific redressing security vulnerabilities. Given the criticality of the component, I would not rely only on unit tests.

File: browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc (hunk 5 above)
Comment (thypon) [reply browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: https://github.com/brave/brave-core/pull/39139#issuecomment-5333844906

Comment (ShivanKaul) [reply browser/ui/views/frame/focus_mode_url_coherence_browsertest.cc]: can this be resolved?

File: browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc [hunk 8]
```diff
@@ -119,3 +131,120 @@ IN_PROC_BROWSER_TEST_F(FocusModeTitleBarViewBrowserTest, TitleBarWiring) {
   browser()->GetFeatures().focus_mode_controller()->SetEnabled(false);
   EXPECT_FALSE(title_bar()->GetVisible());
 }
+
+// Verifies that the Focus Mode title bar's domain label stays coherent with
+// the active page across navigation edge cases (client/server redirects and
+// back/forward cache restore).
+class FocusModeUrlCoherenceBrowserTest
```
Comment (bridiver) [inline browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc]: I think having a separate "coherence" test is a bit confusing. I would drop "coherence" in general and just add these as individual tests focusing on what they actually check, but this still seems unit testable?

File: browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc [hunk 9]
```diff
@@ -119,3 +131,120 @@ IN_PROC_BROWSER_TEST_F(FocusModeTitleBarViewBrowserTest, TitleBarWiring) {
...
+    : public FocusModeTitleBarViewBrowserTest {
+ public:
+  FocusModeUrlCoherenceBrowserTest() {
+    // Enable back/forward cache so the restore test exercises the cached-RFH
```
Comment (bridiver) [inline browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc]: The important thing to explain in comments is why we are doing things and it's not clear to me why we are doing this here?

File: browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc [hunk 10]
```diff
@@ -119,3 +131,120 @@ IN_PROC_BROWSER_TEST_F(FocusModeTitleBarViewBrowserTest, TitleBarWiring) {
...
+// back to the restored page's host. bfcache restore fires DidFinishNavigation;
+// TabUIHelper notifies, and the title bar must re-render with the prior host.
+IN_PROC_BROWSER_TEST_F(FocusModeUrlCoherenceBrowserTest,
+                       BackForwardCacheRestoreUpdatesDomain) {
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(
+      browser(), https_server_.GetURL("a.test", "/empty.html")));
+  ASSERT_TRUE(WaitForDomainToContain(u"a.test"))
+      << base::UTF16ToUTF8(domain_text());
+
+  // Capture the RFH that should enter bfcache when we navigate away. Wrap it
```
Comment (bridiver) [inline browser/ui/views/frame/focus_mode_title_bar_view_browsertest.cc]: I don't understand why we care about this? These internal details shouldn't affect how this works.
